File: leetcode/golang/278.第一个错误版本.go
/** 
 * Forward declaration of isBadVersion API.
 * @param   version   your guess about first bad version
 * @return 	 	      true if current version is bad 
 *			          false if current version is good
 * func isBadVersion(version int) bool;
 */

 func firstBadVersion(n int) int {
    start, end := 1, n;
    for {
        if start >= end {
            break;
        }
        mid := start + (end - start) / 2;
        if isBadVersion(mid) {
            if mid == 1 || !isBadVersion(mid - 1) {
                return mid;
            };
            end = mid - 1;
        } else {
            start = mid + 1;
        }
    };
    if start == end && isBadVersion(start) {
        return start;
    }
    return start;
}

File: leetcode/python/278.第一个错误版本.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
# def isBadVersion(version):

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        start, end = 1, n

        while start < end:

            mid = start + int((end - start) / 2)
            if isBadVersion(mid):
                if mid == 1 or not isBadVersion(mid - 1):
                    return mid
                end = mid - 1
            else:
                start = mid + 1
        
        if start == end and isBadVersion(start):
            return start